mempool/src/sync/mod.rs

```diff
@@ -0,0 +1,421 @@
+pub(crate) mod messages;
+mod tests;
+
+use std::{
+    collections::{
+        hash_map::Entry::{Occupied, Vacant},
+        HashMap, VecDeque,
+    },
+    pin::Pin,
+    sync::Arc,
+    task::{Context, Poll},
+    time::{Duration, Instant},
+};
+
+use futures::{future::BoxFuture, FutureExt, Stream, StreamExt};
+use messages::{
+    MempoolTransactionType, RequestMempoolHashes, RequestMempoolTransactions,
+    ResponseMempoolHashes, ResponseMempoolTransactions,
+};
+use nimiq_blockchain::Blockchain;
+use nimiq_consensus::{sync::syncer::SyncEvent, ConsensusProxy};
+use nimiq_hash::{Blake2bHash, Hash};
+use nimiq_network_interface::{
+    network::Network,
+    peer_info::Services,
+    request::{request_handler, RequestError},
+};
+use nimiq_time::{sleep_until, Sleep};
+use nimiq_transaction::{historic_transaction::RawTransactionHash, Transaction};
+use nimiq_utils::{spawn, stream::FuturesUnordered};
+use parking_lot::RwLock;
+use tokio_stream::wrappers::BroadcastStream;
+
+use crate::{executor::PubsubIdOrPeerId, mempool_state::MempoolState};
+
+enum RequestStatus {
+    Pending,
+    Requested,
+    Received,
+}
+
+/// Struct for keeping track which peers have a specific hash in their mempool and
+/// store if we have a requested the corresponding transaction already to one of those peers.
+struct HashRequestStatus<N: Network> {
+    status: RequestStatus,
+    peer_ids: Vec<N::PeerId>,
```

Comment:
```suggestion
    peer_ids: HashSet<N::PeerId>,
```
Looks like this wants fast lookup, so a `HashSet` would be more appropriate.